okhttp/src/main/java/okhttp3/AsyncDns.java

```diff
@@ -0,0 +1,66 @@
+package okhttp3;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import javax.annotation.Nullable;
+
+public final class AsyncDns implements Dns {
```

Comment:
Not sure how well this will fly given it introduces new public API.  I think this is high impact so would like some solution like this that doesn't involve blocking sync calls for Dns.

I wonder whether the public API part should be limited to a sub-interface with an additional method.